src/core/tsi/s2a/handshaker/s2a_handshaker_client.cc

```diff
@@ -16,41 +16,261 @@
+                             GRPC_SLICE_LENGTH(*bytes_received)));
+  return s2a_get_serialized_session_req(request, arena.ptr());
 }
 
-void s2a_handshaker_client_shutdown(const s2a_handshaker_client* client) {
+tsi_result s2a_handshaker_client::next(grpc_slice* bytes_received) {
+  GPR_ASSERT(bytes_received != nullptr);
+  grpc_slice_unref_internal(recv_bytes_);
+  recv_bytes_ = grpc_slice_ref_internal(*bytes_received);
+  grpc_byte_buffer* buffer = s2a_get_serialized_next(bytes_received);
+  if (buffer == nullptr) {
+    gpr_log(GPR_ERROR, kS2AGetSerializedNextFailed);
+    return TSI_INTERNAL_ERROR;
+  }
+  grpc_byte_buffer_destroy(send_buffer_);
+  send_buffer_ = buffer;
+  tsi_result call_result = make_grpc_call(/*is_start=*/false);
+  if (call_result != TSI_OK) {
+    gpr_log(GPR_ERROR, kS2AMakeGrpcCallFailed);
+  }
+  return call_result;
+}
+
+/** ------------------- Callback methods. -------------------------------- **/
+
+static void s2a_on_status_received(void* arg, grpc_error* error) {
+  // TODO(mattstev): implement.
   return;
 }
 
+/** ------------------- Create, shutdown, & destroy methods. ------------- **/
+
+s2a_handshaker_client::s2a_handshaker_client(
+    s2a_tsi_handshaker* handshaker, grpc_channel* channel,
+    grpc_pollset_set* interested_parties, grpc_s2a_credentials_options* options,
+    const grpc_slice& target_name, grpc_iomgr_cb_func grpc_cb,
+    tsi_handshaker_on_next_done_cb cb, void* user_data, bool is_client) {
+  gpr_mu_init(&mu_);
+  gpr_ref_init(&refs_, 1);
+  grpc_caller_ = grpc_call_start_batch_and_execute;
+  handshaker_ = handshaker;
+  cb_ = cb;
+  user_data_ = user_data;
+  send_buffer_ = nullptr;
+  recv_buffer_ = nullptr;
+  target_name_ = grpc_slice_copy(target_name);
+  if (options != nullptr) {
+    options_ = options->copy();
+  }
+  recv_bytes_ = grpc_empty_slice();
+  grpc_metadata_array_init(&recv_initial_metadata_);
+  is_client_ = is_client;
+  buffer_size_ = kS2AInitialBufferSize;
+  buffer_ = static_cast<uint8_t*>(gpr_zalloc(buffer_size_));
+  handshake_status_details_ = grpc_empty_slice();
+  grpc_slice slice =
+      grpc_slice_from_copied_string(options->handshaker_service_url().c_str());
+  call_ = (options->handshaker_service_url().compare(
+               kS2AHandshakerServiceUrlForTesting) == 0)
+              ? nullptr
+              : grpc_channel_create_pollset_set_call(
+                    channel, /*parent_call=*/nullptr, GRPC_PROPAGATE_DEFAULTS,
+                    interested_parties,
+                    grpc_slice_from_static_string(kS2AServiceMethod), &slice,
+                    GRPC_MILLIS_INF_FUTURE, /*reserved=*/nullptr);
+  GRPC_CLOSURE_INIT(&on_handshaker_service_resp_recv_, grpc_cb, this,
+                    grpc_schedule_on_exec_ctx);
+  GRPC_CLOSURE_INIT(&on_status_received_, s2a_on_status_received, this,
+                    grpc_schedule_on_exec_ctx);
+  grpc_slice_unref_internal(slice);
+}
+
 tsi_result s2a_handshaker_client_create(
```

Comment:
Do we still need this creation API? now that we can directly call the constructor of `s2a_handshaker_client`? It looks more like C style.

Reply:
Per our offline discussion, I'm leaving the create/destroy API's in order to catch error cases, e.g. if the caller has misconfigured the options.

Reply:
SGTM